test/libsinsp_e2e/event_capture.h

```diff
@@ -32,184 +32,133 @@ limitations under the License.
-
-	T* operator->() {
-		if(!m_object_lock.owns_lock()) {
-			throw concurrent_object_handle_state_error(
-			        "Attempt to access wrapped object without obtaining a lock.");
-		}
-		return m_object_ptr;
-	}
-
-	inline T* safe_ptr() { return operator->(); }
-
-	T& operator*() {
-		if(!m_object_lock.owns_lock()) {
-			throw concurrent_object_handle_state_error(
-			        "Attempt to access wrapped object without obtaining a lock.");
-		}
-		return *m_object_ptr;
-	}
-
-	T* unsafe_ptr() { return m_object_ptr; }
-
-	void unlock() { m_object_lock.unlock(); }
-
-private:
-	concurrent_object_handle(sinsp* object_ptr, std::mutex& object_mutex):
-	        m_object_ptr(object_ptr),
-	        m_object_lock(object_mutex, std::defer_lock) {}
-
-	T* m_object_ptr;
-	std::unique_lock<std::mutex> m_object_lock;
-};
+// Just a stupid fake FD value to signal to stop capturing events from driver and exit.
+// Note: we don't use it through eventfd because we want to make sure
+// that we received all events from the driver, until this very last close(FD_SIGNAL_STOP);
+#define EVENTFD_SIGNAL_STOP 1
 
 class callback_param {
 public:
 	sinsp_evt* m_evt;
 	sinsp* m_inspector;
 };
 
-typedef std::function<void(sinsp* inspector)> before_open_t;
-typedef std::function<void(sinsp* inspector)> before_close_t;
+// Right before inspector->start_capture() gets called.
+// Engine is already opened (thus scap handle is already alive).
+typedef std::function<void(sinsp* inspector)> before_capture_t;
+// Right after inspector->stop_capture() gets called.
+// Engine is still opened (thus scap handle is still alive).
+typedef std::function<void(sinsp* inspector)> after_capture_t;
+// Only events matching the filter function are passed to the captured_event_callback_t.
 typedef std::function<bool(sinsp_evt* evt)> event_filter_t;
+// On event callback
 typedef std::function<void(const callback_param& param)> captured_event_callback_t;
-
-// Returns true/false to indicate whether the capture should continue
-// or stop
-typedef std::function<bool()> capture_continue_t;
-
-typedef std::function<void(concurrent_object_handle<sinsp> inspector)> run_callback_t;
+typedef std::function<void(sinsp* inspector)> run_callback_t;
+typedef std::function<void()> run_callback_async_t;
 
 class event_capture {
 public:
-	void init_inspector();
+	event_capture(captured_event_callback_t captured_event_callback,
+	              before_capture_t before_open,
+	              after_capture_t before_close,
+	              event_filter_t filter,
+	              uint32_t max_thread_table_size,
+	              uint64_t thread_timeout_ns,
+	              uint64_t inactive_thread_scan_time_ns);
+
+	void start(bool dump);
+	void stop();
 	void capture();
-	void stop_capture();
-	void wait_for_capture_start();
-	void wait_for_capture_stop();
 
 	static void do_nothing(sinsp* inspector) {}
 
-	static bool always_continue() { return true; }
+	static void run(const run_callback_t& run_function,
```

Comment:
We now have 2 different fashioned `run` static functions exposed by the `event_capture` class: one run the test callback (`run_function`) synchronously and then gathers all generated events; this one can also access the inspector in the callback since it is synchronous.

The other one is async, thus the callback is prevented to access the inspector.

Reply:
Note: i kept the same name because i found it easier to integrate; so the only difference between a sync and an async invocation is just the first parameter type (run_callback_t vs run_callback_async_t).